Add Swarm management backend

tonistiigi opened the pull request:
As described in our [ROADMAP.md](https://github.com/docker/docker/blob/master/ROADMAP.md#14cluster-capable-engine), introduce new Swarm management API endpoints relying on docker/swarmkit to deploy services. It currently vendors [docker/engine-api](docker/engine-api#265) changes.

This PR is fully backward compatible (joining a Swarm is an optional feature of the Engine, and existing commands are not impacted).

Signed-off-by: Tonis Tiigi tonistiigi@gmail.com
Signed-off-by: Victor Vieux vieux@docker.com
Signed-off-by: Daniel Nephin dnephin@docker.com

tonistiigi committed:
Support force new cluster

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Add Swarm management backend

As described in our ROADMAP.md, introduce new Swarm management API
endpoints relying on swarmkit to deploy services. It currently vendors
docker/engine-api changes.

This PR is fully backward compatible (joining a Swarm is an optional
feature of the Engine, and existing commands are not impacted).

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>
Signed-off-by: Victor Vieux <vieux@docker.com>
Signed-off-by: Daniel Nephin <dnephin@docker.com>

tonistiigi committed:
Swarm integration tests

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>
Signed-off-by: Victor Vieux <vieux@docker.com>

dnephin committed:
Add Swarm management CLI commands

As described in our ROADMAP.md, introduce new Swarm management commands
to call to the corresponding API endpoints.

This PR is fully backward compatible (joining a Swarm is an optional
feature of the Engine, and existing commands are not impacted).

Signed-off-by: Daniel Nephin <dnephin@docker.com>
Signed-off-by: Victor Vieux <vieux@docker.com>
Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

Commit:
Add swarm api reference docs

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

amitshukla committed:
CLI docs

Signed-off-by: Amit Shukla <amit.shukla@docker.com>

thaJeztah commented:
Looks like some tests are failing

tonistiigi committed:
Set acceptance policy synchronously in init

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Remove daemon conf dir override

Not needed because no tests using the key
use multiple daemons. Can be added separately.

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

dnephin committed:
New format for setting mounts on a service.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

tonistiigi committed:
Test updates

- Set reconciliation timeout as a variable
- Add timeout to shutdown and increase test timeout to include time needed
for cluster component shutdown.

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

cpuguy83 committed:
Fix panic if daemon is stopped before api is ready

This is easily visibile when the daemon is part of a swarm, and starting
the daemon before raft has finish initializing (which can take some
time).

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

Commit:
Shorten `docker service` restart policy flags

Rename `--restart-policy-*` flags to `--restart-*` (e.g.,
`--restart-policy-delay` becomes `--restart-delay`).

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

Commit:
Clarify services update policies

Rename `--updateconfig-*` to `--update-*` and improve help messages.

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

vieux committed:
use ID for '*'

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
add CACertHash: backend

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
add CACertHash: cli

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
fix panic @vieux introduced

Signed-off-by: Victor Vieux <vieux@docker.com>

Commit:
Minor typo fix

Signed-off-by: Jérôme Petazzoni <jerome.petazzoni@gmail.com>

tonistiigi committed:
Update tests

- Manual acceptance
- Promote/demote
- Raft timeouts

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Update integration to latest swarmkit

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Fix accept policy initialization

- Set secret when using default policy
- Always initialize policy for all roles

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Update join command message

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

dnephin committed:
Fix CLI side of volume problem. (backend)

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
Fix CLI side of volume problem.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

vieux committed:
update to newest engine-api (backend)

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
update to newest engine-api (cli)

Signed-off-by: Victor Vieux <vieux@docker.com>

Commit:
Rename `--port` to `--publish` in `docker service`

Rename `--port` to `--publish` for consistency between `docker service`
and `docker run`.

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

Commit:
Rename `SwarmPort` to `PublicPort`

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

tonistiigi committed:
Remove rename test fix

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Fix RenameCheckNames test

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Add test for CAHash

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

mavenugo committed:
Make swarm-mode incompatible with --cluster-store configurations

In order to take care of init/leave/init cases, we need a reliable
way to reset the ClusterProvider and hence exposed the
SetClusterProvider Daemon API via the backend interface.
With this change, we handle the various combinations of swarm init and
--cluster-store configurations to be reliably detect as incompatible.

Signed-off-by: Madhu Venugopal <madhu@docker.com>

vieux committed:
removed fields from engine-api (backend)

Signed-off-by: Victor Vieux <vieux@docker.com>

tonistiigi committed:
Fix rolling update on upscaling

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

dnephin committed:
Fixes #273 - set command line args in args not command

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
backport fixes from docker/docker PR comments.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

mavenugo committed:
Avoid persisting dynamically created network

Signed-off-by: Madhu Venugopal <madhu@docker.com>

tonistiigi committed:
Fix possible deadlocks

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Skipped stopped container on shutdown

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

vieux committed:
upadate engine-api: backend

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
upadate engine-api: cli

Signed-off-by: Victor Vieux <vieux@docker.com>

tonistiigi committed:
Update integration to new swarmkit

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

vieux committed:
fix compilation

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
fix panic in secret

Signed-off-by: Victor Vieux <vieux@docker.com>

mavenugo committed:
Prevent unmanaged containers to use swarm managed networks

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
accept partial-id for `docker network rm`

Signed-off-by: Madhu Venugopal <madhu@docker.com>

Commit:
Pretty print strings on human-readable outputs

Capitalize first letter in output meant for human consumptions (e.g.,
`docker service ls`, or `docker node inspect --pretty`).

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

Commit:
Fix incorrect `docker service inspect` help message

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

aluzzardi commented on docs/reference/api/docker_remote_api_v1.24.md:
```diff
@@ -3306,6 +3305,1119 @@ Status Codes
+**Warning:** this endpoint is part of the Swarm management feature introduced in Docker 1.12, and
+might be subject to non backward-compatible changes.
+
+`GET /swarm/tasks`
```
Should be `GET /tasks`

/cc @icecrime

tonistiigi committed:
Update integration to new swarmkit

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Fix flakiness on container shutdown

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

mrjana committed:
Recover from stale ingress network plumbing

Currently on cluster ungraceful shutdown ingrss network plumbing can
be left behind and on restart we need to properly cleanup the stale
plumbing to make it useful.

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

mountkin commented on daemon/cluster/convert/swarm.go:
```diff
@@ -0,0 +1,112 @@
+		swarm.Spec.AcceptancePolicy.Policies = append(swarm.Spec.AcceptancePolicy.Policies, types.Policy{
+			Role:       types.NodeRole(policy.Role.String()),
+			Autoaccept: policy.Autoaccept,
+			Secret:     string(policy.Secret.Data),
```
Panicked here when running `docker swarm inspect`.
```
2016-06-12 15:02:57.184744 I | http: panic serving @: runtime error: invalid memory address or nil pointer dereference
goroutine 234 [running]:
net/http.(*conn).serve.func1(0xc8203c5b80)
        /usr/local/go/src/net/http/server.go:1389 +0xc1
panic(0x1a6c400, 0xc820010070)
        /usr/local/go/src/runtime/panic.go:443 +0x4e9
github.com/docker/docker/daemon/cluster/convert.SwarmFromGRPC(0xc820ee0280, 0x19, 0x9, 0xc8208c5e40, 0xc8208c5e50, 0xc8208c5e11, 0x7, 0x0, 0xc8208c5ea0, 0x2, ...)
        /go/src/github.com/docker/docker/daemon/cluster/convert/swarm.go:50 +0x267
github.com/docker/docker/daemon/cluster.(*Cluster).Inspect(0xc8200b9680, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/docker/docker/daemon/cluster/cluster.go:411 +0x23f
github.com/docker/docker/api/server/router/swarm.(*swarmRouter).inspectCluster(0xc82072a7e0, 0x7f350b7d37d8, 0xc82089db30, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0, 0xc82089d6e0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/router/swarm/cluster_routes.go:48 +0x67
github.com/docker/docker/api/server/router/swarm.(*swarmRouter).(github.com/docker/docker/api/server/router/swarm.inspectCluster)-fm(0x7f350b7d37d8, 0xc82089db30, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0, 0xc82089d6e0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/router/swarm/cluster.go:30 +0x74
github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1(0x7f350b7d37d8, 0xc82089db30, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0, 0xc82089d6e0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/middleware/version.go:56 +0x9aa
github.com/docker/docker/api/server/middleware.UserAgentMiddleware.WrapHandler.func1(0x7f350b7d37d8, 0xc82089d980, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0, 0xc82089d6e0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/middleware/user_agent.go:45 +0x68e
github.com/docker/docker/api/server/middleware.DebugRequestMiddleware.func1(0x7f350b7c5940, 0xc8200110f8, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0, 0xc82089d6e0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/middleware/debug.go:21 +0xd4d
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1(0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0)
        /go/src/github.com/docker/docker/api/server/server.go:139 +0x138
net/http.HandlerFunc.ServeHTTP(0xc820859820, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0)
        /usr/local/go/src/net/http/server.go:1618 +0x3a
github.com/gorilla/mux.(*Router).ServeHTTP(0xc82000d0e0, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0)
        /go/src/github.com/docker/docker/vendor/src/github.com/gorilla/mux/mux.go:98 +0x29e
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP(0xc8206c71a0, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0)
        /go/src/github.com/docker/docker/api/server/router_swapper.go:29 +0x72
net/http.serverHandler.ServeHTTP(0xc8202a7f80, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0)
        /usr/local/go/src/net/http/server.go:2081 +0x19e
net/http.(*conn).serve(0xc8203c5b80)
        /usr/local/go/src/net/http/server.go:1472 +0xf2e
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:2137 +0x44e
```

Steps to reproduce:
1. Start docker service with `/usr/bin/dockerd --debug --raw-logs`
2. `docker swarm init`
3. `docker swarm inspect`

vdemeester replied:
Same here :sweat: 

```
root@020824023c36:/go/src/github.com/docker/docker# docker swarm inspect
[]
An error occurred trying to connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.24/swarm: EOF
```

vieux committed:
fix case

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
fix docker swarm inspect output on error

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
git rm reject.go

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
add -h back to inspect

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
cli

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
backend

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
fix mounts

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
api

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
backend

Signed-off-by: Victor Vieux <vieux@docker.com>

dnephin committed:
Upate from PR review.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

tonistiigi committed:
Fix tasks endpoint in reference docs

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

mavenugo committed:
swarm managed networks should me marked as "swarm" scoped (backend)

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
swarm managed networks should me marked as "swarm" scoped (cli)

Signed-off-by: Madhu Venugopal <madhu@docker.com>

Commit:
Add `docker service scale` command

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

mrjana committed:
Vendoring libnetwork @892e1b9864d1f (backend)

Adds overlay datapath security enhancement
Adds netlink performance enhancements
Bug fixes

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

mrjana committed:
Vendoring libnetwork @892e1b9864d1f (cli)

Adds overlay datapath security enhancement
Adds netlink performance enhancements
Bug fixes

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

dongluochen committed:
Upate auto-accept help text.

Signed-off-by: Dong Chen <dongluo.chen@docker.com>

dnephin committed:
Small UX improvements to 'docker service scale'

Signed-off-by: Daniel Nephin <dnephin@docker.com>

cpuguy83 committed:
Fix scale pointer value in service inspect -p

Before this change, pretty output shows `859533555312` instead of `1`.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

Commit:
Add missing newline

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

tonistiigi committed:
Remove state on failed init

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

mrjana committed:
Do not fail on predefined network for managed network

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

dperny committed:
Fix command not displaying in service ls.

Closes #396.

Signed-off-by: Drew Erny <drew.erny@docker.com>

mrjana committed:
Update backend to libnetwork @e8da32ce5693

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

vieux committed:
fix rebase

Signed-off-by: Victor Vieux <vieux@docker.com>

icecrime commented:
LGTM :100:

mavenugo commented:
🎉 🎉 LGTM

vieux commented:
LGTM 💯 this is very good!

icecrime commented:
API reference will be completed in a followup :+1:

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "swarm-backend" git@github.com:docker/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~77
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.
